toolchain/base/int_store.h

```diff
@@ -0,0 +1,422 @@
+  }
+
+  // Construct an ID from a raw 32-bit ID value.
+  static constexpr auto MakeRaw(int32_t raw_id) -> IntId {
+    return IntId(raw_id);
+  }
+
+  // Tests whether the ID is a value ID.
+  //
+  // Only *valid* IDs can have an embedded value, so when true this also implies
+  // the ID is valid.
+  constexpr auto is_value() const -> bool { return id_ > ZeroIndexId; }
+
+  // Tests whether the ID is an index ID.
+  //
+  // Note that an invalid ID is represented as an index ID, so this is *not*
+  // sufficient to test whether an ID is valid.
+  constexpr auto is_index() const -> bool { return id_ <= ZeroIndexId; }
+
+  // Test whether an ID is valid.
+  //
+  // This does not distinguish between value and index IDs, only whether valid.
+  constexpr auto is_valid() const -> bool { return id_ != InvalidId; }
+
+  // Converts an ID to the embedded value. Requires that `is_value()` is true.
+  constexpr auto AsValue() const -> int {
+    CARBON_DCHECK(is_value());
+    return id_;
+  }
+
+  // Converts an ID to an index. Requires that `is_index()` is true.
+  //
+  // Note that this does *not* require that the ID is valid. An invalid ID will
+  // turn into an invalid index.
+  constexpr auto AsIndex() const -> int {
+    CARBON_DCHECK(is_index());
+    return ZeroIndexId - id_;
+  }
+
+  constexpr auto AsTokenPayload() const -> uint32_t {
+    uint32_t payload = id_;
+    // Ensure this ID round trips as the token payload.
+    CARBON_DCHECK(*this == MakeFromTokenPayload(payload));
+    return payload;
+  }
+
+  constexpr auto AsRaw() const -> int32_t { return id_; }
+
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << "int [";
+    if (is_value()) {
+      out << "value: " << AsValue() << "]";
+    } else if (is_index()) {
+      out << "index: " << AsIndex() << "]";
+    } else {
+      CARBON_CHECK(!is_valid());
+      out << "invalid]";
+    }
+  }
+
+  friend constexpr auto operator==(IntId lhs, IntId rhs) -> bool {
+    return lhs.id_ == rhs.id_;
+  }
+  friend constexpr auto operator<=>(IntId lhs, IntId rhs)
+      -> std::strong_ordering {
+    return lhs.id_ <=> rhs.id_;
+  }
+
+ private:
+  friend class IntStore;
+  friend Testing::IntStoreTestPeer;
+
+  // The shift needed when adjusting a between a `TokenIdBits`-width integer and
+  // a 32-bit integer.
+  static constexpr int TokenIdBitsShift = 32 - TokenIdBits;
+
+  // The maximum embedded value in an ID.
+  static constexpr int32_t MaxValue =
+      std::numeric_limits<int32_t>::max() >> TokenIdBitsShift;
+
+  // The ID value that represents an index of `0`. This is the first ID value
+  // representing an index, and all indices are `<=` to this.
+  //
+  // Because this is the first index ID, and we encoded indices as successive
+  // negative numbers counting downwards, we can both use a comparison with
+  // this ID to distinguish value and index IDs, and to compute the actual index
+  // from the ID. The computation of an index in fact is just a subtraction:
```

Comment:
I'm having trouble reading this due to the commas. What do you think of:

```suggestion
  // ZeroIndexId is the first index ID, and we encode indices as successive
  // negative numbers counting downwards. The setup allows us to both use a comparison with
  // this ID to distinguish value and index IDs, and to compute the actual index
  // from the ID.
  //
  // The computation of an index in fact is just a subtraction:
```